pkg/metadata/store/metadata_store.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	"fmt"
+	"sync"
+)
+
+// All byte arrays are expected to be read-only. User MUST NOT modify byte
+// array element directly!!
+
+// UpdateFunc is function used to update a specific metadata. The value
+// passed in is the old value. The updated value should be returned.
+// If there is an error, the update will be rolled back.
+type UpdateFunc func([]byte) ([]byte, error)
```

Comment:
As is comment above:
```
// All byte arrays are expected to be read-only. User MUST NOT modify byte
// array element directly!!
```

Actually, we could pass in a copy of the current data. However, that introduces extra overhead and actually all user will just marshal the byte array without changing it directly.